## Credentials

In [2]:
import pandas as pd
import time 

import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import getpass
import pprint

## GetPass

In [3]:
client_id = str(getpass.getpass('client_id: '))
client_secret = str(getpass.getpass('client_secret: '))

client_id: ········
client_secret: ········


In [5]:
sp = spotipy.Spotify(auth_manager = SpotifyClientCredentials(client_id = client_id, client_secret = client_secret))

## Get the data from a whole playlist

In [6]:
# function which handlees more than 100 results

def get_playlist_tracks(user_id,playlist_id):
    results = sp.user_playlist_tracks(user_id,playlist_id)
    tracks=results["items"]
    while results['next']:
        results = sp.next(results)
        tracks.extend(results['items'])
    return tracks

In [7]:
results2 = get_playlist_tracks("Klaus Tienaho","3PlzvUAGHzXYS4m5fOkuHP")
results2

[{'added_at': '2010-08-19T21:51:01Z',
  'added_by': {'external_urls': {'spotify': 'https://open.spotify.com/user/'},
   'href': 'https://api.spotify.com/v1/users/',
   'id': '',
   'type': 'user',
   'uri': 'spotify:user:'},
  'is_local': False,
  'primary_color': None,
  'track': {'album': {'album_type': 'album',
    'artists': [{'external_urls': {'spotify': 'https://open.spotify.com/artist/5M52tdBnJaKSvOpJGz8mfZ'},
      'href': 'https://api.spotify.com/v1/artists/5M52tdBnJaKSvOpJGz8mfZ',
      'id': '5M52tdBnJaKSvOpJGz8mfZ',
      'name': 'Black Sabbath',
      'type': 'artist',
      'uri': 'spotify:artist:5M52tdBnJaKSvOpJGz8mfZ'}],
    'available_markets': [],
    'external_urls': {'spotify': 'https://open.spotify.com/album/714ndVxSx8lIWhQxdbcXIs'},
    'href': 'https://api.spotify.com/v1/albums/714ndVxSx8lIWhQxdbcXIs',
    'id': '714ndVxSx8lIWhQxdbcXIs',
    'images': [{'height': 640,
      'url': 'https://i.scdn.co/image/ab67616d0000b273cfa6ec6d5374ce8aec1a73f5',
      'width': 

In [8]:
len(results2)

497

## Iterate over the tracks of a playlist to collect the song name, song URI/ID and the artist

In [9]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results2

for r in results2:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf2 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf2['song_name'] = songdf2['song_name'].str.lower()
songdf2['artist'] = songdf2['artist'].str.lower()
songdf2

song_name  \
0                                             paranoid   
1                                        ace of spades   
2                                          crazy train   
3    the number of the beast - 1998 remastered version   
4       hallowed be thy name - 1998 remastered version   
..                                                 ...   
492                                           gangland   
493                                      trial by fire   
494                              whiplash (remastered)   
495                                       rock of ages   
496                                         photograph   

                    song_id              artist  
0    3L8WgDYS5TtwbyxOPSAiFf       black sabbath  
1    47FyQCd3TYLrZ9TU6MPaWK           motörhead  
2    3ao8tE7CIEo3KehDQYYoWv       ozzy osbourne  
3    1s4Ie0cT6P73SRSfh3oyGW         iron maiden  
4    1iGXvUsVVkYBas0Cniw6NB         iron maiden  
..                      ...                 ...  
492  1alav0oB46PY36eKNN6e3X  tygers of pan tang  
493  4NKqJMlONXzac3DEmh5iwp           testament  
494  4k1IS2uF2dGcU3GzOOZqqb           metallica  
495  06VOoCjZ6OfCbOKVIhCIT9         def leppard  
496  5jF3jBoctZU17gY0FFYpVE         def leppard  

[497 rows x 3 columns]

## Get the audio features

In [10]:
feats=sp.audio_features(tracks=song_id[0:99])

In [11]:
feats[0].keys()

dict_keys(['danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'type', 'id', 'uri', 'track_href', 'analysis_url', 'duration_ms', 'time_signature'])

In [12]:
audio_feat_df1=pd.DataFrame(sp.audio_features(tracks=song_id[0:100]))
audio_feat_df2=pd.DataFrame(sp.audio_features(tracks=song_id[100:200]))
audio_feat_df3=pd.DataFrame(sp.audio_features(tracks=song_id[200:300]))
audio_feat_df4=pd.DataFrame(sp.audio_features(tracks=song_id[300:400]))
audio_feat_df5=pd.DataFrame(sp.audio_features(tracks=song_id[400:500]))

In [13]:
audio_feat_df3

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.311   0.806    4   -12.337     1       0.0499      0.091900   
1          0.478   0.937    0    -6.173     0       0.1060      0.007710   
2          0.475   0.576    9   -13.059     1       0.0312      0.021800   
3          0.448   0.833    6    -8.582     1       0.0432      0.004270   
4          0.457   0.779    7    -6.323     1       0.2570      0.018500   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.416   0.792    6   -11.885     1       0.0604      0.003520   
96         0.397   0.969    9    -3.892     1       0.0463      0.000002   
97         0.346   0.765    8    -8.558     1       0.0413      0.002290   
98         0.493   0.938    4    -7.221     0       0.0819      0.000023   
99         0.158   0.973    4    -4.000     0       0.2130      0.000843   

    instrumentalness  liveness  valence    tempo            type  \
0           0.003670    0.1410   0.0772  135.121  audio_features   
1           0.018100    0.8400   0.2940  131.418  audio_features   
2           0.022100    0.3910   0.5240   96.378  audio_features   
3           0.000703    0.2310   0.5920  116.676  audio_features   
4           0.000002    0.0247   0.7340   86.573  audio_features   
..               ...       ...      ...      ...             ...   
95          0.105000    0.4020   0.5200  104.708  audio_features   
96          0.517000    0.0513   0.7020  134.288  audio_features   
97          0.000008    0.2130   0.6440  106.463  audio_features   
98          0.015900    0.1210   0.7070   94.655  audio_features   
99          0.742000    0.0938   0.1370  167.167  audio_features   

                        id                                   uri  \
0   1n5tv95U5Ixc9kEki1C5MA  spotify:track:1n5tv95U5Ixc9kEki1C5MA   
1   1EORu8Coheibl21VjHtbRg  spotify:track:1EORu8Coheibl21VjHtbRg   
2   25suYHVUwWFTG6G31kuhZH  spotify:track:25suYHVUwWFTG6G31kuhZH   
3   2VVat48YTOk3wALkMjNxvQ  spotify:track:2VVat48YTOk3wALkMjNxvQ   
4   0N4NO2BmOIBWrCQ9r8Ei8m  spotify:track:0N4NO2BmOIBWrCQ9r8Ei8m   
..                     ...                                   ...   
95  1bWe6OdgzmJrXLsJN2HDXx  spotify:track:1bWe6OdgzmJrXLsJN2HDXx   
96  2nWjhmRImt6LScZIPQMBnM  spotify:track:2nWjhmRImt6LScZIPQMBnM   
97  3G0EKJZy0j3rMG077UawaC  spotify:track:3G0EKJZy0j3rMG077UawaC   
98  1Iw3wchVrb4rahrtYqcdW6  spotify:track:1Iw3wchVrb4rahrtYqcdW6   
99  0J58LxQpWoigdtTDXrKnX0  spotify:track:0J58LxQpWoigdtTDXrKnX0   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/1n5tv95U5Ixc...   
1   https://api.spotify.com/v1/tracks/1EORu8Coheib...   
2   https://api.spotify.com/v1/tracks/25suYHVUwWFT...   
3   https://api.spotify.com/v1/tracks/2VVat48YTOk3...   
4   https://api.spotify.com/v1/tracks/0N4NO2BmOIBW...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/1bWe6OdgzmJr...   
96  https://api.spotify.com/v1/tracks/2nWjhmRImt6L...   
97  https://api.spotify.com/v1/tracks/3G0EKJZy0j3r...   
98  https://api.spotify.com/v1/tracks/1Iw3wchVrb4r...   
99  https://api.spotify.com/v1/tracks/0J58LxQpWoig...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/1n5t...       639867   
1   https://api.spotify.com/v1/audio-analysis/1EOR...       298401   
2   https://api.spotify.com/v1/audio-analysis/25su...       397773   
3   https://api.spotify.com/v1/audio-analysis/2VVa...       327400   
4   https://api.spotify.com/v1/audio-analysis/0N4N...       313667   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/1bWe...       211733   
96  https://api.spotify.com/v1/audio-analysis/2nWj...       218293   
97  https://api.spotify.com/v1/audio-analysis/3G0E...       284427   
98  https://api.spotify.com/v1/audio-analysis/1Iw3...       146173   
99

In [14]:
frames = [audio_feat_df1, audio_feat_df2,audio_feat_df3, audio_feat_df4,audio_feat_df5]
result = pd.concat(frames)
result.reset_index(drop=True, inplace=True)
result

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0           0.418   0.668    6   -11.297     0       0.0820      0.000143   
1           0.324   0.991    6    -4.465     1       0.1970      0.002470   
2           0.473   0.918    9    -3.580     1       0.0566      0.014500   
3           0.420   0.930    2    -5.583     0       0.0638      0.018900   
4           0.304   0.909    9    -5.162     1       0.0362      0.001850   
..            ...     ...  ...       ...   ...          ...           ...   
492         0.297   0.991    7    -2.365     1       0.2430      0.000106   
493         0.331   0.791    1   -11.187     1       0.0425      0.000257   
494         0.386   0.911    4    -6.474     0       0.0956      0.000608   
495         0.710   0.806   11    -7.660     0       0.0447      0.009470   
496         0.600   0.873    4    -6.542     1       0.0507      0.000354   

     instrumentalness  liveness  valence    tempo            type  \
0             0.00113    0.1730    0.379  162.663  audio_features   
1             0.00895    0.1040    0.195  140.245  audio_features   
2             0.00248    0.3270    0.586  137.606  audio_features   
3             0.00675    0.2640    0.344   99.641  audio_features   
4             0.12200    0.3620    0.348  103.990  audio_features   
..                ...       ...      ...      ...             ...   
492           0.00122    0.2730    0.189  116.428  audio_features   
493           0.42600    0.2100    0.309   94.907  audio_features   
494           0.44300    0.1350    0.188  158.045  audio_features   
495           0.08950    0.1800    0.621  105.866  audio_features   
496           0.42700    0.0604    0.632  123.676  audio_features   

                         id                                   uri  \
0    3L8WgDYS5TtwbyxOPSAiFf  spotify:track:3L8WgDYS5TtwbyxOPSAiFf   
1    47FyQCd3TYLrZ9TU6MPaWK  spotify:track:47FyQCd3TYLrZ9TU6MPaWK   
2    3ao8tE7CIEo3KehDQYYoWv  spotify:track:3ao8tE7CIEo3KehDQYYoWv   
3    1s4Ie0cT6P73SRSfh3oyGW  spotify:track:1s4Ie0cT6P73SRSfh3oyGW   
4    1iGXvUsVVkYBas0Cniw6NB  spotify:track:1iGXvUsVVkYBas0Cniw6NB   
..                      ...                                   ...   
492  1alav0oB46PY36eKNN6e3X  spotify:track:1alav0oB46PY36eKNN6e3X   
493  4NKqJMlONXzac3DEmh5iwp  spotify:track:4NKqJMlONXzac3DEmh5iwp   
494  4k1IS2uF2dGcU3GzOOZqqb  spotify:track:4k1IS2uF2dGcU3GzOOZqqb   
495  06VOoCjZ6OfCbOKVIhCIT9  spotify:track:06VOoCjZ6OfCbOKVIhCIT9   
496  5jF3jBoctZU17gY0FFYpVE  spotify:track:5jF3jBoctZU17gY0FFYpVE   

                                            track_href  \
0    https://api.spotify.com/v1/tracks/3L8WgDYS5Ttw...   
1    https://api.spotify.com/v1/tracks/47FyQCd3TYLr...   
2    https://api.spotify.com/v1/tracks/3ao8tE7CIEo3...   
3    https://api.spotify.com/v1/tracks/1s4Ie0cT6P73...   
4    https://api.spotify.com/v1/tracks/1iGXvUsVVkYB...   
..                                                 ...   
492  https://api.spotify.com/v1/tracks/1alav0oB46PY...   
493  https://api.spotify.com/v1/tracks/4NKqJMlONXza...   
494  https://api.spotify.com/v1/tracks/4k1IS2uF2dGc...   
495  https://api.spotify.com/v1/tracks/06VOoCjZ6OfC...   
496  https://api.spotify.com/v1/tracks/5jF3jBoctZU1...   

                                          analysis_url  duration_ms  \
0    https://api.spotify.com/v1/audio-analysis/3L8W...       167933   
1    https://api.spotify.com/v1/audio-analysis/47Fy...       166707   
2    https://api.spotify.com/v1/audio-analysis/3ao8...       290067   
3    https://api.spotify.com/v1/audio-analysis/1s4I...       291373   
4    https://api.spotify.com/v1/audio-analysis/1iGX...       432800   
..                                                 ...          ...   
492  https://api.spotify.com/v1/audio-analysis/1ala...       220627   
493  https://api.spotify.com/v1/audio-analysis/4NKq...       255360   
494  https://api.spotify.com/v1/audio-analysis/4k1I...       248547   
495  https://ap

## Add more songs

- [80s Heavy Metal](https://open.spotify.com/playlist/7bcOsrWwa5jASp5zwr6RXD)
- by Kjetil Myhre
- 106 songs

In [15]:
results3 = get_playlist_tracks("Kjetil Myhre","7bcOsrWwa5jASp5zwr6RXD")
len(results3)

106

In [16]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results3

for r in results3:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf3 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf3['song_name'] = songdf3['song_name'].str.lower()
songdf3['artist'] = songdf3['artist'].str.lower()
songdf3

song_name                 song_id          artist
0        rainbow in the dark  4Ld00UQYL3aPcK4yOnv0JP             dio
1        girls, girls, girls  5Rso4igx5lTkTAbDEN9nuQ     mötley crüe
2                  red alert  72GcnatH6Q1IqRrNjOuukN      quiet riot
3                     poison  5XcZRgJv3zMhTqCyESjQrF    alice cooper
4        people of the night  6K4fiQRAp9QdwDdK5DKHcl           shout
..                       ...                     ...             ...
101           rock her bones  02m82m2wWh1dzfmkL5mDlX           shout
102        come out and play  3SPJpKR25VMjWwIo2JCL6k  twisted sister
103               seven seas  0vkP5rs4NqGdEW10oHstO8             tnt
104  rock me - 2005 remaster  5RK9Q1vogXWI60qVGDzTZA     great white
105           dream warriors  7AUFwSivoAvJoh1OJDLrGJ          dokken

[106 rows x 3 columns]

-----

- [Heavy Metal-80's Hair Metal-Power Ballads-Rock](https://open.spotify.com/playlist/1tvCfaqcXP2QEHgFWf0t17?si=02bcc9709c514252&nd=1)
- by Solveig Seldal
- 785 songs

In [17]:
results4 = get_playlist_tracks("Solveig Seldal","1tvCfaqcXP2QEHgFWf0t17")
len(results4)

785

In [18]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results4

for r in results4:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf4 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf4['song_name'] = songdf4['song_name'].str.lower()
songdf4['artist'] = songdf4['artist'].str.lower()
songdf4

song_name                 song_id          artist
0    rock you like a hurricane  46QazXxQS0B31CnbRCy8CV       scorpions
1             cartesian dreams  26QQFxxg7EkPLwfAc2qCpy  house of lords
2                 i walk alone  5m8JHIXmasB1qOPCzfGVBr            jorn
3                   she's gone  2fvObZjCzCF57MKmn1ePPO      steelheart
4                slow an' easy  0qTvq8jgrRd2mu3Q7d4oQb      whitesnake
..                         ...                     ...             ...
780          bring me everyone  148V2ZB9jpUeJENStJjjDq        mustasch
781           the skeleton key  6IVKiq8Xxhvg03MUtu5smv           epica
782               head up high  0uRUJf3fZDrTGM9JksGZuZ        firewind
783               lost chapter  207K1MXwebxpXDS43OazOU       pentakill
784              edge of night  3Ro1AMjH9qdY3ZTWfxIFpN       pentakill

[785 rows x 3 columns]

-----

- [80's,rock,heavy metal and metal mix](https://open.spotify.com/playlist/2Rhu5twFe7cNUlRtadJtnw?si=91332b798c8045ca&nd=1)
- by Nicholas Kauffmann
- 668 songs

In [19]:
results5 = get_playlist_tracks("Nicholas Kauffmann","2Rhu5twFe7cNUlRtadJtnw")
len(results5)

668

In [20]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results5

for r in results5:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf5 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf5['song_name'] = songdf5['song_name'].str.lower()
songdf5['artist'] = songdf5['artist'].str.lower()
songdf5

song_name                 song_id                 artist
0                      paranoid  3q9G9SvgdfE1fxlvoDIb20          black sabbath
1            kickstart my heart  7GonnnalI2s19OCQO1J7Tf            mötley crüe
2    bringin' on the heartbreak  5DKSOabxbMTF1oLdPp5bcs            def leppard
3             over the mountain  0i1RTnH2Lj5gTDRU5wtyT2          ozzy osbourne
4                    am i evil?  03pJkRoQU8Lrmgc14ffJGJ              metallica
..                          ...                     ...                    ...
663                     rock me  0u6toDywZABXlUBq76iPKW            great white
664          find your way back  1QlHzR2Nl7gJlb4n3OrtrV     jefferson starship
665                        jane  2RNd6VgxG19db0zs0M1C4Z     jefferson starship
666                  no way out  5ncLNCx0nnz763Aug1q4ay     jefferson starship
667         california dreamin'  1ZEOIhSn6BKErV59bIgn76  the mamas & the papas

[668 rows x 3 columns]

-----

- [Heavy Metal Classics](https://open.spotify.com/playlist/27gN69ebwiJRtXEboL12Ih?si=13cc331c40b04ae0&nd=1)
- by Misi
- 386 songs

In [21]:
results6 = get_playlist_tracks("Misi","27gN69ebwiJRtXEboL12Ih")
len(results6)

386

In [22]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results6

for r in results6:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf6 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf6['song_name'] = songdf6['song_name'].str.lower()
songdf6['artist'] = songdf6['artist'].str.lower()
songdf6

song_name                 song_id  \
0                            paranoid  3q9G9SvgdfE1fxlvoDIb20   
1         the trooper - 2015 remaster  4OROzZUy6gOWN4UGQVaZMF   
2                       ace of spades  47FyQCd3TYLrZ9TU6MPaWK   
3    run to the hills - 2015 remaster  4Zc7TCHzuNwL0AFBlyLdyr   
4    fear of the dark - 2015 remaster  6p8eFfPw3nQkf37aT3AkmK   
..                                ...                     ...   
381                      over my head  54XuDwn7kL2PXGw6vUW6Hx   
382                    wait and bleed  15DLl1r2zi07Ssq5RT1yT0   
383                         powertrip  3azSRGmcZSaSZVV3Izg4Tg   
384                         astronomy  67TDP21NkdDbavSA0sMk8v   
385                            bomber  5FvOGIzAQuJyNaEM62mhIG   

               artist  
0       black sabbath  
1         iron maiden  
2           motörhead  
3         iron maiden  
4         iron maiden  
..                ...  
381          king's x  
382          slipknot  
383    monster magnet  
384  blue öyster cult  
385         motörhead  

[386 rows x 3 columns]

-----

- [80s Heavy Rock / Metal](https://open.spotify.com/playlist/0EPz7rNCG4jjzgHpxdM1nD?si=2acca1ad7dca4c6a&nd=1)
- by rol-ba
- 243 songs

In [23]:
results7 = get_playlist_tracks("rol-ba","0EPz7rNCG4jjzgHpxdM1nD")
len(results7)

243

In [24]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results7

for r in results7:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf7 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf7['song_name'] = songdf7['song_name'].str.lower()
songdf7['artist'] = songdf7['artist'].str.lower()
songdf7

song_name                 song_id  \
0                              breaking the law  2RaA6kIcvomt77qlIgGhCT   
1                                 ace of spades  6EPRKhUOdiFSQwGBRBbvsZ   
2                         cum on feel the noize  3vV3cr2TpPqFk07zxYUbla   
3                                    holy diver  54yLLvpezvPO1lA3GspiZj   
4    run to the hills - 1998 remastered version  2JS1iE5A5RHvUPH5Zl9jlF   
..                                          ...                     ...   
238                                 black metal  3yNoEJifUJdly8ucYoWRwl   
239                                        evil  6LGoJAyN1xAqF3VJZt0MKx   
240                                metal church  34LXW73avo8xz0NLiVuiow   
241                   rock you like a hurricane  58XWGx7KNNkKneHdprcprX   
242                            still loving you  7kjsCbksOZotNNHOvKsJJh   

            artist  
0     judas priest  
1        motörhead  
2       quiet riot  
3              dio  
4      iron maiden  
..             ...  
238          venom  
239  mercyful fate  
240   metal church  
241      scorpions  
242      scorpions  

[243 rows x 3 columns]

-----

- [80s Hard Rock / Heavy Metal](https://open.spotify.com/playlist/0iNXcT6n7EHngUO3JpaW6J?si=4d69f6297b8144c8&nd=1)
- by Rolando Batista
- 547 songs

In [25]:
results8 = get_playlist_tracks("Rolando Batista","0iNXcT6n7EHngUO3JpaW6J")
len(results8)

547

In [26]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results8

for r in results8:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf8 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf8['song_name'] = songdf8['song_name'].str.lower()
songdf8['artist'] = songdf8['artist'].str.lower()
songdf8

song_name  \
0                            rock you like a hurricane   
1                                  sweet child o' mine   
2                    summertime girls - studio version   
3                                                 jump   
4                                       rock the night   
..                                                 ...   
542                                            ingrown   
543                                       united sheep   
544                                      dominoes fall   
545                                 life on your knees   
546  proud to be loud (as heard in great expectations)   

                    song_id              artist  
0    0XIvZ82aDF7JiSi3ZE320u           scorpions  
1    4MhTFsyqIJnjsOweVcU8ug       guns n' roses  
2    2kR3B09M6KeJnchOkxwszt                 y&t  
3    2PfBSkCKZDXMsJogMT0wML           van halen  
4    4NffXDmI70vw19P5Efoe0T              europe  
..                      ...                 ...  
542  1KTQLypZn07OlnGT38efvU            neurosis  
543  6QA6acjmIhVnep0WrinGaV            neurosis  
544  33S9AsrE3wmDzDwoQOzOS1            neurosis  
545  1cnANrdgRpq8IwfS2tBm61            neurosis  
546  4g115GpyzP0U1d7I5RuHJm  dead green mummies  

[547 rows x 3 columns]

-------

- ['80s Metal Classics](https://open.spotify.com/playlist/3KkzYovFLMOiaXrt72xY2d?si=038511fa54a1461c&nd=1)
- by Gonzalo Andres Quintana
- 336 songs

In [29]:
results9 = get_playlist_tracks("Gonzalo Andres Quintana","3KkzYovFLMOiaXrt72xY2d")
len(results9)

336

In [30]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results9

for r in results9:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf9 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf9['song_name'] = songdf9['song_name'].str.lower()
songdf9['artist'] = songdf9['artist'].str.lower()
songdf9

song_name                 song_id         artist
0                      metal gods  0VFfZVCFNe1I2IIChVBR5m   judas priest
1                             one  64Ret7Tf2M8pDE4aqbW2tX      metallica
2                         we rock  5qo8HipgkkuQ9tItiBiTaO            dio
3                riding the storm  4hfpMRkHOW3xGt8zE8ogFk   running wild
4    wasted years - 2015 remaster  1HM7Vp84E4SMzdrZONg6bH    iron maiden
..                            ...                     ...            ...
331            soldier of fortune  6WkCJYM3wtIOZvskzHWYnt       loudness
332        the burning of atlanta  16sznMthzmElmJ4fPFP1aL       whiplash
333         power thrashing death  7x0eznaiq2CaLUfOqiUSvt       whiplash
334                the apparition  3KhTOoJDjpFhcAKUAbMxIB  fates warning
335                      guardian  6OWafYToeRoamwzuIFhtIh  fates warning

[336 rows x 3 columns]

------

- [80s Heavy Metal / Underrated playlist](https://open.spotify.com/playlist/41bkoMrOtBPKTIldGrgTGT?si=c96c2e3f5a654db7&nd=1)
- by ocnarf_
- 606 songs


In [31]:
results10 = get_playlist_tracks("ocnarf_","41bkoMrOtBPKTIldGrgTGT")
len(results10)

606

In [32]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results10

for r in results10:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf10 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf10['song_name'] = songdf10['song_name'].str.lower()
songdf10['artist'] = songdf10['artist'].str.lower()
songdf10

song_name                 song_id            artist
0       wrathchild - 2015 remaster  1SpuDZ7y1W4vaCzHeLvsf7       iron maiden
1                   chaos descends  3vcnNJJNC1g7KoByCaQajL      cirith ungol
2                  dead on arrival  17dwaNuNN2WtEf44tOgVb5       grim reaper
3    i am the one you warned me of  5dvxMjve4VBNYSYkP1VIG3  blue öyster cult
4           rocking the night away  5dKUTtcf4bJvcZ3v5YvnMd       witch cross
..                             ...                     ...               ...
601              fire in the brain  3arsaI4qLpEL9vAX7z6pJl                oz
602                too hot to rock  4Sqe3SNeMajB0bal8GhqyZ         ironcross
603               midwinter nights  0cw0x5fneOxfHUm4QDOwvd             tarot
604               back in the fire  1542KEu4H2dSblQn2iFIrA             tarot
605                         pharao  0jB54mh5HFujdS6nsYpSjb             tarot

[606 rows x 3 columns]

----

- [Women of Heavy Metal](https://open.spotify.com/playlist/2nCbzvyUCzIiWhzxyPKNUO?si=1c87a9ddf7fa4214&nd=1)
- by Michael
- 361 songs


In [33]:
results11 = get_playlist_tracks("Michael","2nCbzvyUCzIiWhzxyPKNUO")
len(results11)

361

In [34]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results11

for r in results11:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf11 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf11['song_name'] = songdf11['song_name'].str.lower()
songdf11['artist'] = songdf11['artist'].str.lower()
songdf11

song_name                 song_id               artist
0                      pisces  1XT5VDdxwI2Bx01MxURMcz               jinjer
1               rule of nines  1Uo51u6JqpIroI98zyhCvE            spiritbox
2    general of the dark army  5WDroVCEVjLayAJoqeuMR7  unleash the archers
3           reason to believe  02gIPDHCTuKgKEl2VlgPzt           arch enemy
4                     silvera  5uunXHE4kIW6uS4HWAXaOQ               gojira
..                        ...                     ...                  ...
356                holy diver  1B4syOWfRxrFYUYn6gieij      burning witches
357            wings of steel  1qNPzB8EXeVCD3O9da9qYD      burning witches
358                spellbound  4DXswc4p8mel7E6C1ZSHDq          lacuna coil
359                      burn  5flMLmYH9APBlSwlAv45ls    we are the fallen
360              never enough  4sIWOzf7LxpuGInB68CrLl                epica

[361 rows x 3 columns]

-----

- [Women singing Heavy Metal](https://open.spotify.com/playlist/3K2G0yetuJDUTJZt5Ey7Oo?si=e1fad9a0f7504e81&nd=1)
- by Fiona
- 81 songs

In [35]:
results12 = get_playlist_tracks("Fiona","3K2G0yetuJDUTJZt5Ey7Oo")
len(results12)

81

In [36]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results12

for r in results12:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf12 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf12['song_name'] = songdf12['song_name'].str.lower()
songdf12['artist'] = songdf12['artist'].str.lower()
songdf12

song_name                 song_id  \
0                     sign of satan  4Jx6PfRbJbFjuWUz6Rd2mt   
1                 (i'm your) victim  7399vKY7Q7QOPBMTNIEZBd   
2                       night curse  1JxaFzaiN8HtrVovhTvxVt   
3             i want you to rock me  6VtP45gz771bs1MTyo02bV   
4                       black widow  35ZnecAxQaTgnoOpRNCbvy   
..                              ...                     ...   
76                     bad attitude  7KkTwKGjGxKjVt52ZBExVJ   
77                         antidote  2U5WbPl1kN2T1F14lY6A6o   
78  master of disaster - remastered  2Iqr4DVuR1b1JHpzBS9LDR   
79             entoptic petroglyphs  2gmZflhQpWMqwUUlI23kaA   
80                 lake of necrosis  56DOvJP7KyCBDz9OECs5eJ   

                artist  
0              warlock  
1           girlschool  
2   sign of the jackal  
3                vixen  
4      burning witches  
..                 ...  
76             hellion  
77              furies  
78        black knight  
79    grendel's sÿster  
80              mystik  

[81 rows x 3 columns]

---

- [Heavy Metal women](https://open.spotify.com/playlist/1Ry5N0xb6gN3qwjHWWiGey?si=f536a22c66e7486d&nd=1)
- by megarel
- 684 songs

In [37]:
results13 = get_playlist_tracks("megarel","1Ry5N0xb6gN3qwjHWWiGey")
len(results13)

684

In [38]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results13

for r in results13:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf13 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf13['song_name'] = songdf13['song_name'].str.lower()
songdf13['artist'] = songdf13['artist'].str.lower()
songdf13

song_name                 song_id  \
0                     the second stone  1hIDjoxsQghoH6KB2NQWD7   
1               the essence of silence  2cNbwuzTf58ZKyFkbUGCCF   
2                    chemical insomnia  5VtnjY7HB9TeOiVvPhxlkr   
3    reverence - living in the heart -  1KiM6LTyDliIxr6fXVLSRz   
4         omen - the ghoulish malady -  63dcMTJd7dlHAKOdxeRU9t   
..                                 ...                     ...   
679         the pallbearer walks alone  4fGxMI4p6dOxXCutg1vrRx   
680                   my sweet mystery  67K3qtJ8UQO7RPIqxEch6a   
681                         until dawn  63JzJHjLkwlqZiCEnPDPkw   
682                the deep & the dark  1bGs0xiAlMdKBpNSOq7Z9F   
683           in the heat of the night  5yxDU6BiwFl1DEp6xRmKGj   

                  artist  
0                  epica  
1                  epica  
2                  epica  
3                  epica  
4                  epica  
..                   ...  
679     the dark element  
680     the dark element  
681             elvellon  
682  visions of atlantis  
683           to/die/for  

[684 rows x 3 columns]

-----

- [Metal women](https://open.spotify.com/playlist/3yE95McccejdsiqLTRTWWN?si=b4b0393a05c34538&nd=1)
- by Neidy Johana Toro
- 80 songs

In [39]:
results14 = get_playlist_tracks("Neidy Johana Toro","3yE95McccejdsiqLTRTWWN")
len(results14)

80

In [40]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results14

for r in results14:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf14 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf14['song_name'] = songdf14['song_name'].str.lower()
songdf14['artist'] = songdf14['artist'].str.lower()
songdf14

song_name                 song_id      artist
0     buried by terror  70Cd40ePlu1brALzwYNn7E      sinaya
1              gravity  5szGWBFHnhflYhFDHwNyyn  once human
2    i speak astronomy  702e8oJJ8fyTnyvEo6VA7u      jinjer
3          war eternal  0WZZENH0kt3O2cBE8q5IRq  arch enemy
4           tourniquet  1mEylG5rq7k9aU3WVW0jGW     cripper
..                 ...                     ...         ...
75            werewolf  5RkXKR2Mer5sD66Hy1PYlk  dehydrated
76           parallels  4vuQZVq89gIEjcL3bTIfbD  dehydrated
77         under ruins  5D14Kvd3GbWhZqKF2tp8MK     nervosa
78        sledgehammer  1ctuI255E9cSkUAxXYG1lP  once human
79  deceiver, deceiver  3jpGc7gcoTyQsrlfQ332Ql  arch enemy

[80 rows x 3 columns]

------

- [More Metal Goddesses / Women of metal / Heavy Queens](https://open.spotify.com/playlist/47T8Fz9uBTqMlxSGULx4J4?si=31e2c6bfa7c042df&nd=1)
- by issok
- 2397 songs

In [41]:
results15 = get_playlist_tracks("issok","47T8Fz9uBTqMlxSGULx4J4")
len(results15)

2387

In [42]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results15

for r in results15:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf15 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf15['song_name'] = songdf15['song_name'].str.lower()
songdf15['artist'] = songdf15['artist'].str.lower()
songdf15

song_name                 song_id  \
0                               voyage of the fallen  28jEzaO3cR4Sm47ipLhNGo   
1                                         unembraced  40QknqrctohMaNrc90QWbS   
2                                     in remembrance  0LG0FaQVYzisTrAKPa4Jin   
3     i'd do anything for love (but i won't do that)  1rjUwi1saJzERBMrNkpNKF   
4                                         ravenheart  3LNPdEhgpgyvR55F0L9eMc   
...                                              ...                     ...   
2382                          in vain - instrumental  6PG3TDO0RfnZSJzdJaY0YU   
2383                        firelight - instrumental  721suJ3bEr7EojcwCFyWrU   
2384                        mad world - instrumental  0Bv4YIlBJnwNpnFrRQCrrj   
2385                     mercy mirror - instrumental  2no5b2crZ3NbvbRrijC3xS   
2386                    trophy hunter - instrumental  2WxOZ2BvDoqIN7KZJkpGDm   

                 artist  
0               xandria  
1               xandria  
2               xandria  
3               xandria  
4               xandria  
...                 ...  
2382  within temptation  
2383  within temptation  
2384  within temptation  
2385  within temptation  
2386  within temptation  

[2387 rows x 3 columns]

-----

- [This is Doro](https://open.spotify.com/playlist/37i9dQZF1DZ06evO3lMIzh)
- by Spotify
- 50 songs

In [43]:
results16 = get_playlist_tracks("Spotify","37i9dQZF1DZ06evO3lMIzh")
len(results16)

50

In [44]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results16

for r in results16:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf16 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf16['song_name'] = songdf16['song_name'].str.lower()
songdf16['artist'] = songdf16['artist'].str.lower()
songdf16

song_name                 song_id  \
0                                           für immer  7GBt0SpACpV3HXad7tMY0U   
1                               dancing with an angel  24V2sAxzW67whN6kbO0vt1   
2                                          lift me up  5FP8EUI8oLePiW8hEfy9BV   
3                                     even angels cry  00CKqQSBhAr2gePvyIzk0P   
4           a dream that cannot be (feat. doro pesch)  0BQO60IlkGyAcSlFGml79G   
5                                          all we are  1hHj5VZDD0FbXQwKpKiPyX   
6                   if i can't have you - no one will  7gTbLA6ksuzbxEbqrhpt50   
7                           egypt (the chains are on)  08IkDWsiDhYtRUlKPxENvk   
8                                  nichts bleibt mehr  3JhCAGwqY1goilCzj3n8EN   
9                                    breaking the law  3UJhB9zmE7dCX8KqH2JEQy   
10                                      all for metal  7ECsTXBu0NsHrXte3PP36s   
11                                     it still hurts  4G0UdoJ120BHNZukephLWg   
12                            walking with the angels  3rcQfO8DzCmLC9TMQQXalG   
13                              legacy (intro) - live  4WM5ai9PWzpGvrzm4QYicj   
14                                         1000 years  7rSCibeWoBJoOG9JKszcWs   
15                    love me in black - 2020 version  7ff1Img4ScXOJi0Mfu6Qih   
16                             a whiter shade of pale  7B5OsMidvkRxfINzbRBWYC   
17                               kiss of death - live  68mzM2vgKfPQCEus9HL2DM   
18  where the wild wolves have gone (feat. doro pe...  14JVBviJgcqmGxfO7KlhL6   
19                                  angel in the dark  1sPCeOvh7dCAOtJTPhMWi9   
20                                 tausend mal gelebt  2ullWuR07XoTwOyNWszo7M   
21                               touch of evil - live  2DFHkEyG7RQzHor67eBfEP   
22                                           hoffnung  1W1GLa4EYqKhqXSBHle6NQ   
23                                        lonely wolf  2YFIxC7ykNvuNsz7eYwBQl   
24                                let love rain on me  7kgoLZugeoJclVxMLF7r9i   
25                                             salome  1MqDnOphHpSXX1mVK8Sv4U   
26                                      alles ist gut  0hZyPlMIj6oznMEJvcEYaz   
27                                         black rose  4m6r4M964oZpp2G0I7dEsR   
28                                        metal tango  3sglfxqLAhq9yeF5jpYepf   
29                                i know you by heart  6584fVke0wvDnZANC8KQQp   
30                                  fall for me again  5fpNMsoQYfSa8xlIDkwkAI   
31                                          brickwall  7aBYCpaludr9EVl2Nq9E9A   
32                               nothing else matters  5pZIdQ3eQLJDR7LGbNko5f   
33                         the alliance of hellhoundz  5tCcJVscUnLOssG1NnEHjA   
34             llueva en mi tu amor - spanish version  2ntUdwe20dITVK3aPlT5pu   
35                                      bis aufs blut  5aa8AkY5aM1Wm2HKuTJiDa   
36                                 i'll be holding on  2s77tcXn7sL5Fd3DFkyXaX   
37                         raise your fist in the air  00QqqiF4EYMH6JhRvQNYRS   
38                                 ein stück ewigkeit  42VbptVP9ebAsxelWsG9lt   
39                             east meets west - live  2x8AZTI6gLXEpAUhxjdLeK   
40         in freiheit stirbt mein herz - single edit  5EjTC8TopomBOAxtNxmENC   
41                                   beyond the trees  7nJRyVrB671C8Vsfpd8qEW   
42                                             helden  6pEfLtjimY7WfDDKbSvtxE   
43                                       black ballad  7dJ3dETeHaoRMHVRWe8gbN   
44                            cold, cold world - live  5l7TJXavZVRpwcyfDhY6nu   
45                                    world gone wild  3MCv5kY2WB7GOlbsHZxmkS   
46                                            seelied  0URpWr1h979000HTnEG6Hg   
47                                backstage to heaven  66kTuff9bzg8uzPjW1Q438   
48                          alone again - bonus track  5TCC99LnrBfpsTTDP9eq

-----

- [Metal BIG 25](https://open.spotify.com/playlist/2SLuYNdsCmBRYkNYjL8MXj)
- by edgar.salas
- 1008 songs

In [45]:
results17 = get_playlist_tracks("edgar.salas","2SLuYNdsCmBRYkNYjL8MXj")
len(results17)

1008

In [46]:
# collect data with a loop
song_name=[]
song_id=[]
artist=[]
playlist=results17

for r in results17:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf17 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf17['song_name'] = songdf17['song_name'].str.lower()
songdf17['artist'] = songdf17['artist'].str.lower()
songdf17

song_name                 song_id                    artist
0              god of light  1zw5RlIr77z3IdiBfxoXjr                    danzig
1           suicidal maniac  4msqNBGHuOZ59LOjGZITvV       suicidal tendencies
2                    mother  5JtuIHVfsYBFAVrHwjUddj                    danzig
3                am i demon  2Bt7cSAgiPWpM2dOU6VAg1                    danzig
4          expendable youth  2pdvKfAvuwNHGG2y7vMC1y                    slayer
...                     ...                     ...                       ...
1003             down rodeo  40L8L7344pPMCjlZc19RQU  rage against the machine
1004         without a face  3DdOsUq1M9mJBFVR0xbSOx  rage against the machine
1005             roll right  5tYosD9TicWDYO3kAE4ADp  rage against the machine
1006  year of tha boomerang  69oLqnnqggg6v7JMw9VIbf  rage against the machine
1007      persona non grata  12HenFSS0P7hgIznVBxsde                    exodus

[1008 rows x 3 columns]

-----

- [True 80s Heavy Metal](https://open.spotify.com/playlist/4hwXRbhEmxzhAxKBNThKMX?si=a70ef3a321a845e4&nd=1)
- by guitaromnivore
- 1670 songs


In [47]:
results18 = get_playlist_tracks("guitaromnivore","4hwXRbhEmxzhAxKBNThKMX")
len(results18)

1676

In [48]:
song_name=[]
song_id=[]
artist=[]
playlist=results18

for r in results18:
    song_name.append(r['track']['name'])
    song_id.append(r['track']['id'])
    artist.append(r['track']['artists'][0]['name'])

songdf18 = pd.DataFrame({"song_name":song_name, "song_id":song_id, "artist":artist} )
songdf18['song_name'] = songdf18['song_name'].str.lower()
songdf18['artist'] = songdf18['artist'].str.lower()
songdf18

song_name                 song_id          artist
0                       crystal logic  1bP2StEGAk87l0YYoifag8    manilla road
1                             forever  2jE86fMXxqYh5xDEvl1PsP             y&t
2     heaven and hell - 2008 remaster  3Jl5GohfNwozDmpzmQBLDI   black sabbath
3                       kyrie eleison  5PBSbTueCPoqAmFq1rRcKr   fates warning
4                phantom of the opera  33hy20OHQwMFhDlkaDsi6s     iron maiden
...                               ...                     ...             ...
1671                        rock city  2Hd5Q2ioGnvXqBBhWVi35D          krokus
1672                  break my chains  6gSlPytOoUVOfIcJAryGjS          killer
1673                      heatstrokes  26cJQf78xAff9wTp5VmuVW          krokus
1674                     eat the rich  09jtIFItoNKnC86zlzBZ29          krokus
1675                    eye's of fire  3GIVW0jDF7h9fltBpMiioG  medieval steel

[1676 rows x 3 columns]

## Concat all playlists

songdf2 - songdf13

In [49]:
songdf1 = [songdf2,songdf3,songdf4,songdf5,songdf6,songdf7,songdf8,songdf9,songdf10,songdf11,songdf12,songdf13,songdf14,songdf15,songdf16,songdf17]
songdf = pd.concat(songdf1)
songdf.reset_index(drop=True, inplace=True)
songdf

song_name  \
0                                              paranoid   
1                                         ace of spades   
2                                           crazy train   
3     the number of the beast - 1998 remastered version   
4        hallowed be thy name - 1998 remastered version   
...                                                 ...   
8820                                         down rodeo   
8821                                     without a face   
8822                                         roll right   
8823                              year of tha boomerang   
8824                                  persona non grata   

                     song_id                    artist  
0     3L8WgDYS5TtwbyxOPSAiFf             black sabbath  
1     47FyQCd3TYLrZ9TU6MPaWK                 motörhead  
2     3ao8tE7CIEo3KehDQYYoWv             ozzy osbourne  
3     1s4Ie0cT6P73SRSfh3oyGW               iron maiden  
4     1iGXvUsVVkYBas0Cniw6NB               iron maiden  
...                      ...                       ...  
8820  40L8L7344pPMCjlZc19RQU  rage against the machine  
8821  3DdOsUq1M9mJBFVR0xbSOx  rage against the machine  
8822  5tYosD9TicWDYO3kAE4ADp  rage against the machine  
8823  69oLqnnqggg6v7JMw9VIbf  rage against the machine  
8824  12HenFSS0P7hgIznVBxsde                    exodus  

[8825 rows x 3 columns]

In [50]:
# how many unique songs are in the df?
songdf['song_id'].nunique()

7524

In [51]:
# keep fast dont mantain any of the duplicates, last or first mantain one of them
songdf = songdf.drop_duplicates(subset='song_id', keep="last")
songdf.reset_index(drop=True, inplace=True)
songdf

song_name                 song_id                    artist
0                  paranoid  3L8WgDYS5TtwbyxOPSAiFf             black sabbath
1               crazy train  3ao8tE7CIEo3KehDQYYoWv             ozzy osbourne
2                  war pigs  1mnXT8zNssefzrRkeuD8Ov             black sabbath
3             back in black  6SPOlvJok6QmNWkvUVPQMs                     ac/dc
4                  iron man  5TFyNdCLlC1ZRuwczqaS97             black sabbath
...                     ...                     ...                       ...
7519             down rodeo  40L8L7344pPMCjlZc19RQU  rage against the machine
7520         without a face  3DdOsUq1M9mJBFVR0xbSOx  rage against the machine
7521             roll right  5tYosD9TicWDYO3kAE4ADp  rage against the machine
7522  year of tha boomerang  69oLqnnqggg6v7JMw9VIbf  rage against the machine
7523      persona non grata  12HenFSS0P7hgIznVBxsde                    exodus

[7524 rows x 3 columns]

In [52]:
songdf['song_id'].nunique()

7524

## Get the audio features

In [65]:
songdf.rename(columns={'song_id': 'id'},inplace=True, errors='raise')

KeyError: "['song_id'] not found in axis"

In [67]:
songdf

song_name                      id                    artist
0                  paranoid  3L8WgDYS5TtwbyxOPSAiFf             black sabbath
1               crazy train  3ao8tE7CIEo3KehDQYYoWv             ozzy osbourne
2                  war pigs  1mnXT8zNssefzrRkeuD8Ov             black sabbath
3             back in black  6SPOlvJok6QmNWkvUVPQMs                     ac/dc
4                  iron man  5TFyNdCLlC1ZRuwczqaS97             black sabbath
...                     ...                     ...                       ...
7519             down rodeo  40L8L7344pPMCjlZc19RQU  rage against the machine
7520         without a face  3DdOsUq1M9mJBFVR0xbSOx  rage against the machine
7521             roll right  5tYosD9TicWDYO3kAE4ADp  rage against the machine
7522  year of tha boomerang  69oLqnnqggg6v7JMw9VIbf  rage against the machine
7523      persona non grata  12HenFSS0P7hgIznVBxsde                    exodus

[7524 rows x 3 columns]

In [69]:
import numpy as np

In [71]:
features = []
for n, chunk_series in songdf.groupby(np.arange(len(songdf)) // 50).id:
    features += sp.audio_features([*map(str, chunk_series)])
features_df = pd.DataFrame.from_dict(filter(None, features))
tracks_with_features_df = songdf.merge(features_df, on=['id'], how='inner')

In [72]:
tracks_with_features_df

song_name                      id                    artist  \
0                  paranoid  3L8WgDYS5TtwbyxOPSAiFf             black sabbath   
1               crazy train  3ao8tE7CIEo3KehDQYYoWv             ozzy osbourne   
2                  war pigs  1mnXT8zNssefzrRkeuD8Ov             black sabbath   
3             back in black  6SPOlvJok6QmNWkvUVPQMs                     ac/dc   
4                  iron man  5TFyNdCLlC1ZRuwczqaS97             black sabbath   
...                     ...                     ...                       ...   
7519             down rodeo  40L8L7344pPMCjlZc19RQU  rage against the machine   
7520         without a face  3DdOsUq1M9mJBFVR0xbSOx  rage against the machine   
7521             roll right  5tYosD9TicWDYO3kAE4ADp  rage against the machine   
7522  year of tha boomerang  69oLqnnqggg6v7JMw9VIbf  rage against the machine   
7523      persona non grata  12HenFSS0P7hgIznVBxsde                    exodus   

      danceability  energy  key  loudness  mode  speechiness  acousticness  \
0            0.418   0.668    6   -11.297     0       0.0820      0.000143   
1            0.473   0.918    9    -3.580     1       0.0566      0.014500   
2            0.452   0.701    9    -9.534     1       0.0677      0.145000   
3            0.436   0.710    9    -5.666     1       0.0350      0.011300   
4            0.339   0.824    9    -9.522     1       0.0938      0.005210   
...            ...     ...  ...       ...   ...          ...           ...   
7519         0.427   0.867    7    -9.077     1       0.1280      0.005350   
7520         0.425   0.808    9    -8.376     0       0.2210      0.005340   
7521         0.527   0.789    9    -8.049     1       0.0694      0.014000   
7522         0.354   0.711    4   -10.134     0       0.1490      0.000499   
7523         0.455   0.975    2    -6.137     1       0.0822      0.000022   

      instrumentalness  liveness  valence    tempo            type  \
0             0.001130    0.1730    0.379  162.663  audio_features   
1             0.002480    0.3270    0.586  137.606  audio_features   
2             0.008170    0.2670    0.519   90.157  audio_features   
3             0.008110    0.0767    0.735   92.887  audio_features   
4             0.004320    0.0803    0.538  154.573  audio_features   
...                ...       ...      ...      ...             ...   
7519          0.000846    0.0785    0.560  170.256  audio_features   
7520          0.179000    0.1750    0.369   92.257  audio_features   
7521          0.013300    0.3810    0.452   86.137  audio_features   
7522          0.000423    0.2450    0.518   84.268  audio_features   
7523          0.041500    0.3090    0.294  128.377  audio_features   

                                       uri  \
0     spotify:track:3L8WgDYS5TtwbyxOPSAiFf   
1     spotify:track:3ao8tE7CIEo3KehDQYYoWv   
2     spotify:track:1mnXT8zNssefzrRkeuD8Ov   
3     spotify:track:6SPOlvJok6QmNWkvUVPQMs   
4     spotify:track:5TFyNdCLlC1ZRuwczqaS97   
...                                    ...   
7519  spotify:track:40L8L7344pPMCjlZc19RQU   
7520  spotify:track:3DdOsUq1M9mJBFVR0xbSOx   
7521  spotify:track:5tYosD9TicWDYO3kAE4ADp   
7522  spotify:track:69oLqnnqggg6v7JMw9VIbf   
7523  spotify:track:12HenFSS0P7hgIznVBxsde   

                                             track_href  \
0     https://api.spotify.com/v1/tracks/3L8WgDYS5Ttw...   
1     https://api.spotify.com/v1/tracks/3ao8tE7CIEo3...   
2     https://api.spotify.com/v1/tracks/1mnXT8zNssef...   
3     https://api.spotify.com/v1/tracks/6SPOlvJok6Qm...   
4     https://api.spotify.com/v1/tracks/5TFyNdCLlC1Z...   
...                                                 ...   
7519  https://api.spotify.com/v1/tracks/40L8L7344pPM...   
7520  https://api.spotify.com/v1/tracks/3DdOsUq1M9mJ...   
7521  https://api.spotify.com/v1/tracks/5tYosD9TicWD...   
7522  https://api.spotify.com/v1/tracks/69oLqnnqggg6...   
7523  https://api.spotify.com/v1/tracks/12HenFSS0P7h...   

     

In [73]:
#store it in a csv
tracks_with_features_df.to_csv(r'tracks_with_features_df.csv', index = False, header = True)

## APPENDIX - TRASH CODE

In [ ]:
# after this there are a lot of problems but I keep the code in the case i need something later.
# this code is saved as a text cell to not alter the current notebook
# The code below now is a mess because I have change and try to solve the problems but I only destroyed the code :(

In [53]:
# feats=sp.audio_features(tracks=song_id[0:99])

# feats[0].keys()

dict_keys(['danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'type', 'id', 'uri', 'track_href', 'analysis_url', 'duration_ms', 'time_signature'])

In [54]:
# len(song_id)

1676

###### slow way - without iteration (wishlist: iterate it to do it faster and not dying typing :) 
audio_feat_df1=pd.DataFrame(sp.audio_features(tracks=song_id[0:100]))
audio_feat_df2=pd.DataFrame(sp.audio_features(tracks=song_id[100:200]))
audio_feat_df3=pd.DataFrame(sp.audio_features(tracks=song_id[200:300]))
audio_feat_df4=pd.DataFrame(sp.audio_features(tracks=song_id[300:400]))
audio_feat_df5=pd.DataFrame(sp.audio_features(tracks=song_id[400:500]))
audio_feat_df6=pd.DataFrame(sp.audio_features(tracks=song_id[500:600]))
audio_feat_df7=pd.DataFrame(sp.audio_features(tracks=song_id[600:700]))
audio_feat_df8=pd.DataFrame(sp.audio_features(tracks=song_id[700:800]))
audio_feat_df9=pd.DataFrame(sp.audio_features(tracks=song_id[800:900]))
audio_feat_df10=pd.DataFrame(sp.audio_features(tracks=song_id[900:1000]))
audio_feat_df11=pd.DataFrame(sp.audio_features(tracks=song_id[1000:1100]))
audio_feat_df12=pd.DataFrame(sp.audio_features(tracks=song_id[1100:1200]))
audio_feat_df13=pd.DataFrame(sp.audio_features(tracks=song_id[1200:1300]))
audio_feat_df14=pd.DataFrame(sp.audio_features(tracks=song_id[1300:1400]))
audio_feat_df15=pd.DataFrame(sp.audio_features(tracks=song_id[1400:1500]))
audio_feat_df16=pd.DataFrame(sp.audio_features(tracks=song_id[1500:1600]))


audio_feat_df17=pd.DataFrame(sp.audio_features(tracks=song_id[1600:1700]))
audio_feat_df18=pd.DataFrame(sp.audio_features(tracks=song_id[1700:1800]))
audio_feat_df19=pd.DataFrame(sp.audio_features(tracks=song_id[1800:1900]))
audio_feat_df20=pd.DataFrame(sp.audio_features(tracks=song_id[1900:2000]))
audio_feat_df21=pd.DataFrame(sp.audio_features(tracks=song_id[2000:2100]))
audio_feat_df22=pd.DataFrame(sp.audio_features(tracks=song_id[2100:2200]))
audio_feat_df23=pd.DataFrame(sp.audio_features(tracks=song_id[2200:2300]))
audio_feat_df24=pd.DataFrame(sp.audio_features(tracks=song_id[2300:2400]))
audio_feat_df25=pd.DataFrame(sp.audio_features(tracks=song_id[2400:2500]))
audio_feat_df26=pd.DataFrame(sp.audio_features(tracks=song_id[2500:2600]))
audio_feat_df27=pd.DataFrame(sp.audio_features(tracks=song_id[2600:2700]))
audio_feat_df28=pd.DataFrame(sp.audio_features(tracks=song_id[2700:2800]))
audio_feat_df29=pd.DataFrame(sp.audio_features(tracks=song_id[2800:2900]))
audio_feat_df30=pd.DataFrame(sp.audio_features(tracks=song_id[2900:3000]))
audio_feat_df31=pd.DataFrame(sp.audio_features(tracks=song_id[3000:3100]))
audio_feat_df32=pd.DataFrame(sp.audio_features(tracks=song_id[3100:3200]))
audio_feat_df33=pd.DataFrame(sp.audio_features(tracks=song_id[3200:3300]))
audio_feat_df34=pd.DataFrame(sp.audio_features(tracks=song_id[3300:3400]))
audio_feat_df35=pd.DataFrame(sp.audio_features(tracks=song_id[3400:3500]))
audio_feat_df36=pd.DataFrame(sp.audio_features(tracks=song_id[3500:3600]))
audio_feat_df37=pd.DataFrame(sp.audio_features(tracks=song_id[3600:3700]))
audio_feat_df38=pd.DataFrame(sp.audio_features(tracks=song_id[3700:3800]))
audio_feat_df39=pd.DataFrame(sp.audio_features(tracks=song_id[3800:3900]))
audio_feat_df40=pd.DataFrame(sp.audio_features(tracks=song_id[3900:4000]))

In [57]:
# after the df17something happens  
# audio_feat_df17

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.359   0.915    1    -8.090     0       0.0838      0.000004   
1          0.183   0.900    6   -11.531     1       0.0954      0.006060   
2          0.231   0.889    4    -9.340     1       0.0976      0.002820   
3          0.246   0.927   11    -9.494     1       0.0737      0.012100   
4          0.161   0.858    5    -9.245     1       0.0726      0.005140   
..           ...     ...  ...       ...   ...          ...           ...   
71         0.606   0.848    7   -12.034     1       0.0417      0.003970   
72         0.492   0.816    7    -7.821     1       0.0367      0.000011   
73         0.365   0.687    9   -12.666     1       0.0401      0.000638   
74         0.509   0.796    2   -11.728     1       0.0359      0.000480   
75         0.411   0.916    3    -3.921     0       0.0575      0.000017   

    instrumentalness  liveness  valence    tempo            type  \
0           0.005590    0.3180    0.178  164.820  audio_features   
1           0.344000    0.3090    0.384  176.833  audio_features   
2           0.001380    0.0983    0.436  115.977  audio_features   
3           0.003850    0.1160    0.414  109.690  audio_features   
4           0.000006    0.4700    0.518  171.901  audio_features   
..               ...       ...      ...      ...             ...   
71          0.430000    0.2900    0.461  124.612  audio_features   
72          0.839000    0.1510    0.540  110.121  audio_features   
73          0.133000    0.0889    0.336  155.243  audio_features   
74          0.460000    0.1610    0.652  143.664  audio_features   
75          0.000219    0.1120    0.273  120.033  audio_features   

                        id                                   uri  \
0   5e2mcN8BzSDPd0UQH2D1JQ  spotify:track:5e2mcN8BzSDPd0UQH2D1JQ   
1   1RnB4Mc5qAUR2WnTcpTgZD  spotify:track:1RnB4Mc5qAUR2WnTcpTgZD   
2   5w5OfdAbEiz19Z1e5UjB5f  spotify:track:5w5OfdAbEiz19Z1e5UjB5f   
3   5N6nGguJZ6I1OLXphmvJVp  spotify:track:5N6nGguJZ6I1OLXphmvJVp   
4   0ag9PlrOhD3uaizAQUXFYx  spotify:track:0ag9PlrOhD3uaizAQUXFYx   
..                     ...                                   ...   
71  2Hd5Q2ioGnvXqBBhWVi35D  spotify:track:2Hd5Q2ioGnvXqBBhWVi35D   
72  6gSlPytOoUVOfIcJAryGjS  spotify:track:6gSlPytOoUVOfIcJAryGjS   
73  26cJQf78xAff9wTp5VmuVW  spotify:track:26cJQf78xAff9wTp5VmuVW   
74  09jtIFItoNKnC86zlzBZ29  spotify:track:09jtIFItoNKnC86zlzBZ29   
75  3GIVW0jDF7h9fltBpMiioG  spotify:track:3GIVW0jDF7h9fltBpMiioG   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/5e2mcN8BzSDP...   
1   https://api.spotify.com/v1/tracks/1RnB4Mc5qAUR...   
2   https://api.spotify.com/v1/tracks/5w5OfdAbEiz1...   
3   https://api.spotify.com/v1/tracks/5N6nGguJZ6I1...   
4   https://api.spotify.com/v1/tracks/0ag9PlrOhD3u...   
..                                                ...   
71  https://api.spotify.com/v1/tracks/2Hd5Q2ioGnvX...   
72  https://api.spotify.com/v1/tracks/6gSlPytOoUVO...   
73  https://api.spotify.com/v1/tracks/26cJQf78xAff...   
74  https://api.spotify.com/v1/tracks/09jtIFItoNKn...   
75  https://api.spotify.com/v1/tracks/3GIVW0jDF7h9...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/5e2m...       275080   
1   https://api.spotify.com/v1/audio-analysis/1RnB...       243541   
2   https://api.spotify.com/v1/audio-analysis/5w5O...       320440   
3   https://api.spotify.com/v1/audio-analysis/5N6n...       394560   
4   https://api.spotify.com/v1/audio-analysis/0ag9...       274613   
..                                                ...          ...   
71  https://api.spotify.com/v1/audio-analysis/2Hd5...       287627   
72  https://api.spotify.com/v1/audio-analysis/6gSl...       281600   
73  https://api.spotify.com/v1/audio-analysis/26cJ...       240667   
74  https://api.spotify.com/v1/audio-analysis/09jt...       254840   
75

audio_feat_df41=pd.DataFrame(sp.audio_features(tracks=song_id[4000:4100]))
audio_feat_df42=pd.DataFrame(sp.audio_features(tracks=song_id[4100:4200]))
audio_feat_df43=pd.DataFrame(sp.audio_features(tracks=song_id[4200:4300]))
audio_feat_df44=pd.DataFrame(sp.audio_features(tracks=song_id[4300:4400]))
audio_feat_df45=pd.DataFrame(sp.audio_features(tracks=song_id[4400:4500]))
audio_feat_df46=pd.DataFrame(sp.audio_features(tracks=song_id[4500:4600]))
audio_feat_df47=pd.DataFrame(sp.audio_features(tracks=song_id[4600:4700]))
audio_feat_df48=pd.DataFrame(sp.audio_features(tracks=song_id[4700:4800]))
audio_feat_df49=pd.DataFrame(sp.audio_features(tracks=song_id[4800:4900]))
audio_feat_df50=pd.DataFrame(sp.audio_features(tracks=song_id[4900:5000]))
audio_feat_df51=pd.DataFrame(sp.audio_features(tracks=song_id[5000:5100]))
audio_feat_df52=pd.DataFrame(sp.audio_features(tracks=song_id[5100:5200]))
audio_feat_df53=pd.DataFrame(sp.audio_features(tracks=song_id[5200:5300]))
audio_feat_df54=pd.DataFrame(sp.audio_features(tracks=song_id[5300:5400]))
audio_feat_df55=pd.DataFrame(sp.audio_features(tracks=song_id[5400:5500]))
audio_feat_df56=pd.DataFrame(sp.audio_features(tracks=song_id[5500:5600]))
audio_feat_df57=pd.DataFrame(sp.audio_features(tracks=song_id[5600:5700]))
audio_feat_df58=pd.DataFrame(sp.audio_features(tracks=song_id[5700:5800]))
audio_feat_df59=pd.DataFrame(sp.audio_features(tracks=song_id[5800:5900]))
audio_feat_df60=pd.DataFrame(sp.audio_features(tracks=song_id[5900:6000]))
audio_feat_df61=pd.DataFrame(sp.audio_features(tracks=song_id[6000:6100]))
audio_feat_df62=pd.DataFrame(sp.audio_features(tracks=song_id[6100:6200]))
audio_feat_df63=pd.DataFrame(sp.audio_features(tracks=song_id[6200:6300]))
audio_feat_df64=pd.DataFrame(sp.audio_features(tracks=song_id[6300:6400]))
audio_feat_df65=pd.DataFrame(sp.audio_features(tracks=song_id[6400:6500]))
audio_feat_df66=pd.DataFrame(sp.audio_features(tracks=song_id[6500:6600]))
audio_feat_df67=pd.DataFrame(sp.audio_features(tracks=song_id[6600:6700]))
audio_feat_df68=pd.DataFrame(sp.audio_features(tracks=song_id[6700:6800]))
audio_feat_df69=pd.DataFrame(sp.audio_features(tracks=song_id[6800:6900]))
audio_feat_df70=pd.DataFrame(sp.audio_features(tracks=song_id[6900:7000]))
audio_feat_df71=pd.DataFrame(sp.audio_features(tracks=song_id[7000:7100]))
audio_feat_df72=pd.DataFrame(sp.audio_features(tracks=song_id[7100:7200]))
audio_feat_df73=pd.DataFrame(sp.audio_features(tracks=song_id[7200:7300]))
audio_feat_df74=pd.DataFrame(sp.audio_features(tracks=song_id[7300:7400]))
audio_feat_df75=pd.DataFrame(sp.audio_features(tracks=song_id[7400:7500]))
audio_feat_df76=pd.DataFrame(sp.audio_features(tracks=song_id[7500:7523]))


In [ ]:
# frames = [audio_feat_df1,audio_feat_df2,audio_feat_df3,audio_feat_df4,audio_feat_df5,audio_feat_df6,audio_feat_df7,audio_feat_df8,audio_feat_df9,audio_feat_df10,
          # audio_feat_df11,audio_feat_df12,audio_feat_df13,audio_feat_df14,audio_feat_df15,audio_feat_df16,audio_feat_df17,audio_feat_df18,audio_feat_df19,audio_feat_df20,
          # audio_feat_df21,audio_feat_df22,audio_feat_df23,audio_feat_df24,audio_feat_df25, audio_feat_df26,audio_feat_df27,audio_feat_df28,audio_feat_df29,audio_feat_df30,
          # audio_feat_df31,audio_feat_df32,audio_feat_df33,audio_feat_df34,audio_feat_df35,audio_feat_df36,audio_feat_df37,audio_feat_df38,audio_feat_df39,audio_feat_df40,
          # audio_feat_df41,audio_feat_df42,audio_feat_df43,audio_feat_df44,audio_feat_df45,audio_feat_df46,audio_feat_df47,audio_feat_df48,audio_feat_df49,audio_feat_df50,
          # audio_feat_df51,audio_feat_df52,audio_feat_df53,audio_feat_df54,audio_feat_df55,audio_feat_df56,audio_feat_df57,audio_feat_df58,audio_feat_df59,audio_feat_df60,
          # audio_feat_df61,audio_feat_df62,audio_feat_df63,audio_feat_df64,audio_feat_df65,audio_feat_df66,audio_feat_df67, audio_feat_df68,audio_feat_df69,audio_feat_df70,
          # audio_feat_df71,audio_feat_df72,audio_feat_df73,audio_feat_df74,audio_feat_df75,audio_feat_df76]

# audio_features = pd.concat(frames)
# audio_features.reset_index(drop=True, inplace=True)

In [60]:
# audio_features

0  acousticness                                       analysis_url  \
0      NaN      0.000014  https://api.spotify.com/v1/audio-analysis/1bP2...   
1      NaN      0.000366  https://api.spotify.com/v1/audio-analysis/2jE8...   
2      NaN      0.023900  https://api.spotify.com/v1/audio-analysis/3Jl5...   
3      NaN      0.004210  https://api.spotify.com/v1/audio-analysis/5PBS...   
4      NaN      0.000141  https://api.spotify.com/v1/audio-analysis/33hy...   
...    ...           ...                                                ...   
1730  None           NaN                                                NaN   
1731  None           NaN                                                NaN   
1732  None           NaN                                                NaN   
1733  None           NaN                                                NaN   
1734  None           NaN                                                NaN   

      danceability  duration_ms  energy                      id  \
0            0.236     361840.0   0.928  1bP2StEGAk87l0YYoifag8   
1            0.249     345933.0   0.908  2jE86fMXxqYh5xDEvl1PsP   
2            0.329     419933.0   0.673  3Jl5GohfNwozDmpzmQBLDI   
3            0.275     324893.0   0.943  5PBSbTueCPoqAmFq1rRcKr   
4            0.320     426373.0   0.879  33hy20OHQwMFhDlkaDsi6s   
...            ...          ...     ...                     ...   
1730           NaN          NaN     NaN                     NaN   
1731           NaN          NaN     NaN                     NaN   
1732           NaN          NaN     NaN                     NaN   
1733           NaN          NaN     NaN                     NaN   
1734           NaN          NaN     NaN                     NaN   

      instrumentalness  key  liveness  loudness  mode  speechiness    tempo  \
0             0.036800  4.0    0.1250    -7.411   0.0       0.2730  134.188   
1             0.075300  7.0    0.0786    -5.126   0.0       0.0628  154.573   
2             0.109000  3.0    0.1380    -8.465   0.0       0.0296   91.122   
3             0.000004  7.0    0.0458    -8.487   1.0       0.1850  130.452   
4             0.117000  4.0    0.5200    -5.685   0.0       0.0643  109.570   
...                ...  ...       ...       ...   ...          ...      ...   
1730               NaN  NaN       NaN       NaN   NaN          NaN      NaN   
1731               NaN  NaN       NaN       NaN   NaN          NaN      NaN   
1732               NaN  NaN       NaN       NaN   NaN          NaN      NaN   
1733               NaN  NaN       NaN       NaN   NaN          NaN      NaN   
1734               NaN  NaN       NaN       NaN   NaN          NaN      NaN   

      time_signature                                         track_href  \
0                3.0  https://api.spotify.com/v1/tracks/1bP2StEGAk87...   
1                4.0  https://api.spotify.com/v1/tracks/2jE86fMXxqYh...   
2                4.0  https://api.spotify.com/v1/tracks/3Jl5GohfNwoz...   
3                4.0  https://api.spotify.com/v1/tracks/5PBSbTueCPoq...   
4                4.0  https://api.spotify.com/v1/tracks/33hy20OHQwMF...   
...              ...                                                ...   
1730             NaN                                                NaN   
1731             NaN                                                NaN   
1732             NaN                                                NaN   
1733             NaN                                                NaN   
1734             NaN                                                NaN   

                type                                   uri  valence  
0     audio_features  spotify:track:1bP2StEGAk87l0YYoifag8    0.245  
1     audio_features  spotify:track:2jE86fMXxqYh5xDEvl1PsP    0.422  
2     audio_features  spotify:track:3Jl5GohfNwozDmpzmQBLDI    0.323  
3     audio_features  spotify:track:5PBSbTueCPoqAmFq1rRcKr    0.112  
4     audio_features  spotify:track:33hy20OHQwMFhDlkaDs

In [ ]:
# frames = [audio_feat_df1, audio_feat_df2,audio_feat_df3, audio_feat_df4,audio_feat_df5,audio_feat_df6,audio_feat_df7, audio_feat_df8,audio_feat_df9,audio_feat_df10]
# audio_features = pd.concat(frames)
# audio_features.reset_index(drop=True, inplace=True)
# audio_features.rename(columns={'id': 'song_id'},inplace=True, errors='raise')
# audio_features

In [ ]:
# framesb = [audio_feat_df1,audio_feat_df2,audio_feat_df3,audio_feat_df4,audio_feat_df5,audio_feat_df6,audio_feat_df7,audio_feat_df8,audio_feat_df9,audio_feat_df10,
          audio_feat_df11,audio_feat_df12,audio_feat_df13,audio_feat_df14,audio_feat_df15,audio_feat_df16,audio_feat_df17]

# audio_featuresb = pd.concat(framesb)
# audio_featuresb.reset_index(drop=True, inplace=True)
# audio_featuresb

## Merge audio features df and songs df

In [ ]:
# songdf.nunique()

In [ ]:
# audio_featuresb.nunique()

In [ ]:
# songdf.rename(columns={'song_id': 'id_song'},inplace=True, errors='raise')
# audio_features.rename(columns={'id': 'id_song'},inplace=True, errors='raise')

In [ ]:
# songdf

In [ ]:
# song_feat_df = pd.merge(songdf, audio_features, on='id_song') #here song_id is common column
# song_feat_df

# lot of problems to solve later / tomorrow 

# WHYYYYY ONLYYY 287 ROOOOWS!!! 🔫🔫🔫🔫🔫🔫🔫

## Try to optimaze the process 

def getTrackFeatures(id):
  meta = sp.track(id)
  features = sp.audio_features(id)

  ##### meta
  name = meta['name']
  album = meta['album']['name']
  artist = meta['album']['artists'][0]['name']
  release_date = meta['album']['release_date']
  length = meta['duration_ms']
  popularity = meta['popularity']

  ###### features
  acousticness = features[0]['acousticness']
  danceability = features[0]['danceability']
  energy = features[0]['energy']
  instrumentalness = features[0]['instrumentalness']
  liveness = features[0]['liveness']
  loudness = features[0]['loudness']
  speechiness = features[0]['speechiness']
  tempo = features[0]['tempo']
  time_signature = features[0]['time_signature']

  track = [name, album, artist, release_date, length, popularity, danceability, acousticness, danceability, energy, instrumentalness, liveness, loudness, speechiness, tempo, time_signature]
  return track

# loop over track ids 
tracks = []
for song in range(len(songdf)):
  time.sleep(.5)
  track = getTrackFeatures(songdf[song])
  tracks.append(track)

# create dataset
# df = pd.DataFrame(tracks, columns = ['name', 'album', 'artist', 'release_date', 'length', 'popularity', 'danceability', 'acousticness', 'danceability', 'energy', 'instrumentalness', 'liveness', 'loudness', 'speechiness', 'tempo', 'time_signature'])
#df.to_csv("spotify.csv", sep = ',')

def analyzeplaylistdict(playlist_dict):
    
    # Loop through every playlist in the dict and analyze it
    for i, (key, val) in enumerate(playlist_dict.items()):
        playlist_df = analyze_playlist(*val)
        # Add a playlist column so that we can see which playlist a track belongs too
        playlist_df["playlist"] = key
        # Create or concat df
        if i == 0:
            playlist_dict_df = playlist_df
        else:
            playlist_dict_df = pd.concat([playlist_dict_df, playlist_df], ignore_index = True)
            
    return playlist_dict_df

playlist_dict = {
 "results2" : ("Klaus Tienaho","3PlzvUAGHzXYS4m5fOkuHP"), 
 "results3" : ("Kjetil Myhre","7bcOsrWwa5jASp5zwr6RXD"),
 "results4" : ("Solveig Seldal","1tvCfaqcXP2QEHgFWf0t17"),
 "results5" : ("Nicholas Kauffmann","2Rhu5twFe7cNUlRtadJtnw"), 
 "results6" : ("Misi","27gN69ebwiJRtXEboL12Ih"),
 "results7" : ("rol-ba","0EPz7rNCG4jjzgHpxdM1nD"),
 "results8" : ("Rolando Batista","0iNXcT6n7EHngUO3JpaW6J"), 
 "results9" : ("Gonzalo Andres Quintana","3KkzYovFLMOiaXrt72xY2d"),
 "results10" : ("ocnarf_","41bkoMrOtBPKTIldGrgTGT"),
 "results11" : ("Michael","2nCbzvyUCzIiWhzxyPKNUO"), 
 "results12" : ("Fiona","3K2G0yetuJDUTJZt5Ey7Oo"),
 "results13" : ("megarel","1Ry5N0xb6gN3qwjHWWiGey"),
 "results14" : ("Neidy Johana Toro","3yE95McccejdsiqLTRTWWN"),
 "results15" : ("issok","47T8Fz9uBTqMlxSGULx4J4"),
 "results16" : ("Spotify","37i9dQZF1DZ06evO3lMIzh"), 
 "results17" : ("edgar.salas","2SLuYNdsCmBRYkNYjL8MXj"),
 "results18" : ("guitaromnivore","4hwXRbhEmxzhAxKBNThKMX")
}

multiple_playlist_df = analyze_playlist_dict(playlist_dict)
multiple_playlist_df["playlist"].value_counts()

**Next steps**
- review the second prototype definition
- work on completing song collection (green box) including audio features
- develop the second prototype on top of your mvp
- test interaction with the user
- (live) ping spotify to look for the audio features of the not hot song
- (live) billboard scraping